### ISSUE TITLE
Support UnifiedPush

### ISSUE THREAD

**Issue by hockeymikey** [home-assistant/android#3174]
**Is your feature request related to a problem? Please describe.**
[UnifiedPush](https://unifiedpush.org/) is an open source push notification protocol that is implemented by a few apps such as NextPush or ntfysh. After installing one of these implementations (called distributors), you can select them for using them for push notifications in apps such as Element or other [apps](https://unifiedpush.org/users/apps/).
**Describe the solution you'd like**
I'd like to be able to use UnifiedPush for my notifications in the app. This would allow users to have push notifications but not rely on FCM and Google. 

**Describe alternatives you've considered, if any**
There is none.

**Additional context**
I think the links above to their website is a good starting point for understanding the protocol. For implementation on the app side, I think going into one of the apps that support unifiedpush like Element would be a good place to draw inspiration. Here is their [issue ticket for element ](https://github.com/vector-im/element-android/issues/2743)to see their discussion and PRs for the implementation.

**Comment by dshokouhi** [comment]
This needs to first be supported in HA core before the app can consider it. I think we should hold off on this request until then.

**Comment by dshokouhi** [comment]
Let's close this until core has its own request and it gets implemented until then the app can't support this.

**Comment by hockeymikey** [comment]
@dshokouhi But why close it? It's good to track rather than having to duplicate it.

**Comment by hockeymikey** [comment]
https://community.home-assistant.io/t/add-unifiedpush-support-to-home-assistant-android-core/446344

**Comment by julianfoad** [comment]
S1m of the UnifiedPush developers, in [UnifiedPush matrix room](https://matrix.to/#/#unifiedpush:matrix.org), says (paraphrased):

A server supporting webpush is enough to support UnifiedPush. It looks like the HA core supports webpush, so the issue can be reopened :)

There is example code in the [UP-Example](https://github.com/UnifiedPush/android-example) app, for how to get UnifiedPush notifications with webpush.

There are currently open discussions to update the U.P. spec to make it clearly compatible with webpush. ([SERV_1.2.0 #37](https://github.com/UnifiedPush/specifications/pull/37), [Retransmission #34](https://github.com/UnifiedPush/specifications/issues/34)).

Terminology:
  UnifiedPush | Webpush
  ---|---
  End-user app | user agent
  push server | push service
  endpoint | push resources

**Comment by dshokouhi** [comment]
unfortunately its still not that simple, the current `mobile_app` integration is still dependent on firebase and if not using local push will default to doing things like rate limiting notifications. Considerations will need to be made in the integration before the app should support it.

**Comment by schklom** [comment]
@dshokouhi I'm far from an expert, but it sounds to me like there is a misunderstanding.
Implementing UP does not mean getting rid of Firebase. UP can fallback to it

https://unifiedpush.org/developers/embedded_fcm/
> If you want your app to fallback to FCM on Android if:
> 
> 1. Your user doesn’t have any UnifiedPush Distributors and
> 2. They have Google Services enabled
> 
> This is for you. You can embed an FCM Distributor, and if the user doesn’t have another distributor, this one will be used. These libraries basically act like UnifiedPush Distributors, but are internal to the app and pass notifications through FCM.
>
> There are 2 libraries doing it: one using the Google Firebase library and another (beta) entirely FOSS[*](https://en.wikipedia.org/wiki/FOSS) that doesn’t.
>
> ## About
> 1. The [Android Embedded FCM Distributor](https://github.com/UnifiedPush/android-embedded_fcm_distributor) is the original embedded distributor. It uses the official Google library for FCM support. However, it is proprietary thus cannot be included on app-stores like F-Droid.

**Comment by Venefilyn** [comment]
@dshokouhi just checking if you saw the above. I do not want to use Google and this would be a huge bonus for me

Given that it wouldn't he a replacement it seems very possible with the current implementation. It also means it could support F-Droid better as UnifiedPush wouldn't have to be stripped from that distribution

**Comment by Neon-44** [comment]
@dshokouhi 

can you please reopen this or mark it as "not planned"?

It is apparently supported in Core and possible with FCM, so the original reason for closing it doesn't stand anymore.
I understand if you don't have the resources/motivation to do this, but if that is so, please tell us straight instead of hiding behind this pseudo-issue :)

All I want is you to look at it and communicate your conclusion with us, because leaving it in this pseudo-dead / pseudo-alive status is very frustrating :)

**Comment by Handrail9** [comment]
> @dshokouhi
> 
> can you please reopen this or mark it as "not planned"?

Correct me if I'm wrong but it was closed as "Not Planned". I also would love for this feature to happen but I believe you either made a typo or were not clear in what you were asking for.

**Comment by p1gp1g** [comment]
> > @dshokouhi
> > can you please reopen this or mark it as "not planned"?
> 
> Correct me if I'm wrong but it was closed as "Not Planned". I also would love for this feature to happen but I believe you either made a typo or were not clear in what you were asking for.

I think this is because it has been closed with the following message, while the core already has it implemented. It means it has been closed for the wrong reasons :)

> Let's close this until core has its own request and it gets implemented until then the app can't support this.

**Comment by Venefilyn** [comment]
Can you link to the Core now having UP support?

**Comment by p1gp1g** [comment]
Here you are: https://github.com/home-assistant/core/blob/834a1ed608fd015cb68f43c7e0f58cf98f769366/homeassistant/components/html5/notify.py#L487

We are updating UnifiedPush specifications to make UnifiedPush endpoints explicitly web push endpoints. (see https://codeberg.org/UnifiedPush/specifications/pulls/1), it will be merge in the coming month. UnifiedPush is already compatible with it.

**Comment by jpelgrom** [comment]
There seems to be some confusion about where changes are needed.

- Home Assistant core (the server, sending the message) supports any number of notification services.
- The Home Assistant Android app (this repository) specifically uses a notification service built for the apps, which is intended to require no user setup and uses a direct WS connection between the app and server or FCM (via a cloud function hosted by Nabu Casa).

> Here you are: https://github.com/home-assistant/core/blob/834a1ed608fd015cb68f43c7e0f58cf98f769366/homeassistant/components/html5/notify.py#L487

This relates to the first option I pointed out, generic notification services. You don't need the app for this. This also requires more user setup.

> the current `mobile_app` integration is still dependent on firebase and if not using local push will default to doing things like rate limiting notifications

This is what was pointed out previously and is still somewhat valid. The app can tell the server where to send notifications, but the server expects it to be the function previously mentioned. That function is also responsible for the rate limiting. 

I think providing a UP endpoint instead should just work, but we need to manipulate some data in the app when received to have it behave the same, which requires testing. I'm willing to try make that work in the coming weeks.

**Comment by p1gp1g** [comment]
> > Here you are: https://github.com/home-assistant/core/blob/834a1ed608fd015cb68f43c7e0f58cf98f769366/homeassistant/components/html5/notify.py#L487
> 
> This relates to the first option I pointed out, generic notification services. You don't need the app for this. This also requires more user setup.

This function is used for the web application (https://github.com/home-assistant/frontend/blob/dev/src/components/ha-push-notifications-toggle.ts#L69). The idea is to use it with the android application. The android application get the endpoint after registering to UnifiedPush (when `onNewEndpoint` is called, cf https://unifiedpush.org/developers/android/), it sends this endpoint to the core application. And the core application use it. There is no need for user setup (except if they want to self host their server)

This is already used by many applications that way

**Comment by p1gp1g** [comment]
So, to resume (except if I'm wrong):
- there is no change needed on the core side
- there is no need for any user setup
- this will be transparent for the user (like FCM, but without FCM)

**Comment by jpelgrom** [comment]
To use the web push integration, [you need to set it up on the server](https://www.home-assistant.io/integrations/html5/#requirements). This is also a different integration and will create a different `notify` entity than the one created by the app. Just like setting up something like Telegram will create a different `notify` entity.

As mentioned, I think there is a way to make it work with the existing setup (*not* web push) but it requires testing which I'll try doing in the coming weeks.

**Comment by p1gp1g** [comment]
The section "Configuring the platform" where one add their VAPID public key to firebase console is only for chrome browsers to accept the VAPID key. This is not necessary for Firefox browsers and isn't for UnifiedPush too (if the feature is use with UnifiedPush).

For UnifiedPush, the vapid key will be able to be passed during registration (on android, with the coming specifications), cf https://codeberg.org/UnifiedPush/specifications/src/commit/e07dab3deec2f7778188717df3ad1c2a7245334f/specifications/android.md?display=source#L148 , at this moment it is just not checked.

Like for the web app notifications, that's probably ok to say
> If you want UnifiedPush, you need to:
> - install pywebpush (like it is mentioned in the quoted page)
> - and configure it:
>
> ```
> # Example configuration.yaml entry
> notify:
>   - platform: html5
>     vapid_pub_key: YOUR_PUBLIC_KEY
>     vapid_prv_key: YOUR_PRIVATE_KEY
>     vapid_email: YOUR_EMAIL

**Comment by 0verEngineer** [comment]
@dshokouhi please reopen this issue because only the app needs to implement it, the server is already able to use it.

**Comment by hardwareadictos** [comment]
This would be awesome for using companion app on GrapheneOS, or for people who wants de degoogle their devices!! Already used by Element clients for example working well!

**Comment by kuflierl** [comment]
I would lile to bump this issue since it hasn't receved any focus in the last few months... Whats the current state of integrating unified push?

**Comment by p1gp1g** [comment]
The recent blog post on UnifiedPush website may give additional argument: https://unifiedpush.org/news/20250131_push_for_decentralized/

It shows how decentralized project can get push notifications, even for users using Google Services (only), without any gateway, nor anything to do on the firebase console

**Comment by applesoff** [comment]
Any info on this? Would love to decrease the battery drain from home assistant and improve notifications.

**Comment by Mywk** [comment]
Bumping this as having such feature is getting more important nowadays. @ilar already highlighted most of it. Please consider re-opening this @dshokouhi

**Comment by Neon-44** [comment]
I honestly think that @dshokouhi has just muted this thread.

@dshokouhi can you just quickly confirm that you didn't mute this thread and consciously choose to keep it closed?
Because I'm contemplating opening a new Issue to get your attention if you muted it.

**Comment by dshokouhi** [comment]
hey everyone we have not forgotten about this issue. In fact we have been waiting for the results of Joris' research from his earlier comment. The comment that immediately followed his seemed to have gone back to what we were trying to convey that the HTML5 integration has nothing to do with the mobile app integration. I can see several responses after that thinking that the comment was correct but that is not the case. Any mobile app that wants to connect with HA core must use the mobile app integration and not anything else, HA core will reject any PR trying to get around that rule. 

Furthermore the toggle that was requested to work would require not only HA core development but also HA frontend development as the toggle is only for HTML5 integration. Notice how its hidden from the companion app when you open that screen using the app, thats on purpose and predates the android app.

Please keep in mind that while the request is closed that does not prevent anyone from doing their own research and submitting a PR to add functionality if you have done the research to see how things will need to be added. Please also try not to tag Joris asking for updates, that would probably only delay responses. I believe one thing that Joris had discovered during his research was that the app may first need to add support for completely removing notifications before we can let a user switch to Unified Push or anything else.

**Comment by p1gp1g** [comment]
UnifiedPush may be introduced into Firefox, and will most probably be in Fennec, so you should be able to get home-assistant push notifications with the web app

**Comment by NicholasFlamy** [comment]
> Please keep in mind that while the request is closed that does not prevent anyone from doing their own research and submitting a PR to add functionality if you have done the research to see how things will need to be added.

So why don't you just leave it open?

**Comment by Rudd-O** [comment]
Core already has UnifiedPush support using the `ntfy` integration.

Can this please be reopened now?

I've left a comment offering to fund this development here: https://discord.com/channels/330944238910963714/1284965926336335993/threads/1454856120664588368

**Comment by Rudd-O** [comment]
Related

https://github.com/home-assistant/android/issues/3206

**Comment by Rudd-O** [comment]
> Any mobile app that wants to connect with HA core must use the mobile app integration and not anything else, HA core will reject any PR trying to get around that rule.

This isn't entirely correct.  HA Companion already supports receiving notification data from FCM, so clearly HA core devs have not rejected this.  All we're asking is that HA Companion also supports receiving notification data from NTFY (an app which works almost the same as FCM -- it relays notifications to apps on the same device).

Here is what needs to happen for everything to work just right:

* User must install NTFY app on the phone.  He can optionally use his own push server (although the default works quite well).  There are other apps, such as Nextcloud Push, which offer the same interface.
* When user enables the much-awaited UnifiedPush setting in Companion, Companion asks NTFY (or whatever other UP app on the phone) to create a topic, register it on the notification server, and return the endpoint to the Companion app on success.  Companion saves this endpoint locally.
* The Companion app asks HA Core to relay messages via the saved NTFY endpoint.  HA Core registers this endpoint associated with the app.  This mechanism already exists because that's exactly how FCM notifications work.
* When HA Core wants to send a notification to the app, _instead of using the FCM notification endpoint_, it simply uses the NTFY endpoint that Companion gave to it.
* NTFY receives the message through the registered topic, and then relays the message (using intents I presume) to the Companion app.
* The Companion app receives the message, and passes the message to the code that processes FCM-based or WebSocket-based notifications.

That's all that needs to happen.  That's the whole spec.  The defaults don't even need to change.  Most of the code that is required to make this work already exists; it's almost certain that a Companion refactor to make this happen will involve creating a few interfaces, factoring out existing FCM support code to an implementation, and implementing anything UnifiedPush-specific as another implementation of the same interface (it could even be assumed that the implementation for WebSocket and the implementation for UP can be the exact same).  On the server what changes is solely _which HTTP endpoint_ gets the `POST`.

This already exists in the ElementX Matrix chat app.  When the user selects notifications be delivered through UnifiedPush, the phone shows a list of UP apps (usually NTFY or Nextcloud Push) to use.  When the user chooses one of these apps, the exchange to register an ElementX-exclusive endpoint happens, and then ElementX saves the endpoint and tells the Matrix server "please send any and all notifications of type X, Y and Z to http : //...endpoint.../...".  From then on, notifications just work, and the app can happily sleep in between messages.

**Comment by Rudd-O** [comment]
Heck there is even a draft PR https://github.com/home-assistant/android/pull/5261

**Comment by Rudd-O** [comment]
I have finally debugged why the WebSocket method of getting notifications is so battery-destroying.

The client sends a WebSocket ping every 30 seconds, which the server responds to.  Of course, this measure is in place to ensure that any NAT between the client and the server stays up, but I have zero clue as to why 30 seconds were chosen for this traffic.  On Linux, the NAT timeout for established TCP sessions is:

```
 user  ~  cat /proc/sys/net/netfilter/nf_conntrack_tcp_timeout_established 
432000
# (value in seconds)
```

The advantage of this approach is that the connection is kept alive even under the most hostile network conditions.

The disadvantage is that it keeps the mobile device awake for up to a few seconds every 30 seconds.

Which is the thing that destroys the battery, slowly but surely.

My first suggestion is to default this to 15 minutes or more.  The other suggestion is to allow the user to configure this to whatever timeout he decides, with a link to documentation explaining the behavior and configuration of various frontend proxies that are in common use with Home Assistant (e.g. [this](https://nginx.org/en/docs/http/ngx_http_proxy_module.html#proxy_send_timeout) or [this](https://nginx.org/en/docs/http/websocket.html)).

**Comment by NicholasFlamy** [comment]
> I have finally debugged why the WebSocket method of getting notifications is so battery-destroying.
> 
> The client sends a WebSocket ping every 30 seconds, which the server responds to. Of course, this measure is in place to ensure that any NAT between the client and the server stays up, but I have zero clue as to why 30 seconds were chosen for this traffic. On Linux, the NAT timeout for established TCP sessions is:
> 
> ```
>  user  ~  cat /proc/sys/net/netfilter/nf_conntrack_tcp_timeout_established 
> 432000
> # (value in seconds)
> ```
> 
> The advantage of this approach is that the connection is kept alive even under the most hostile network conditions.
> 
> The disadvantage is that it keeps the mobile device awake for up to a few seconds every 30 seconds.
> 
> Which is the thing that destroys the battery, slowly but surely.
> 
> My first suggestion is to default this to 15 minutes or more. The other suggestion is to allow the user to configure this to whatever timeout he decides, with a link to documentation explaining the behavior and configuration of various frontend proxies that are in common use with Home Assistant (e.g. [this](https://nginx.org/en/docs/http/ngx_http_proxy_module.html#proxy_send_timeout)).

Dang, that's a pretty crazy finding. Whoever set it up wasn't thinking about real-world. They were probably thinking about their personal use case.

It could also be someone didn't know how WebSockets worked and though that was setting a 30 second polling without WebSockets.

**Comment by Rudd-O** [comment]
They probably tried various timeouts with a shitty middlebox (perhaps a $20 router?) and found out that 30 seconds worked for them.

Actually, anything above 59 seconds (with NginX frontend-proxying HA at least) will by default cause NginX to close the connection because HA will send nothing for 60 seconds, triggering `proxy_read_timeout`.

https://nginx.org/en/docs/http/websocket.html explains why.  Note that `proxy_send_timeout` is probably going to need tweaking too as well as `send_timeout`.  NginX is not a very good WebSocket proxy, come to think of it.

At home I would be more than okay with the client sending a WebSocket ping at most every 30 minutes.  That should prolong battery life to days.

**Comment by Rudd-O** [comment]
Interestingly there appears to be a way to set a timeout conditionally.

```
    map $http_upgrade $connection_upgrade {
        default upgrade;
        '' close;
    }
    map $http_upgrade $proxy_read_timeout {
        default 86400;
        '' 65;
    }

    server {
        proxy_read_timeout $proxy_read_timeout;
        proxy_http_version 1.1;
        proxy_set_header Upgrade $http_upgrade;
        proxy_set_header Connection $connection_upgrade;
        location / {
            proxy_pass http://backend;
       }
    }
```

**Comment by Rudd-O** [comment]
@NicholasFlamy AHA!

https://square.github.io/okhttp/changelogs/changelog_3x/#version-3100

The recommended ping interval in the library used by HA Companion Android is 30 seconds!

The Companion app however does not set any default ping interval:

https://github.com/home-assistant/android/blob/71ef9597890cd738f314bfac41e9db2fa5c6c65d/common/src/main/kotlin/io/homeassistant/companion/android/common/data/websocket/impl/WebSocketCoreImpl.kt#L193

According to the docs for okHttp 5.x, there will be no pings if a value is not set using pingInterval() (which I couldn't find in the codebase for the Android app).

So it's a mystery where these pings are coming from.  My understanding of the code indicates quite clearly that there shouldn't be any websocket pings.

**Comment by Rudd-O** [comment]
BINGO!

They rolled their own ping implementation!

https://github.com/home-assistant/android/blob/71ef9597890cd738f314bfac41e9db2fa5c6c65d/app/src/main/kotlin/io/homeassistant/companion/android/websocket/WebsocketManager.kt#L171

God knows why.

Here it is!

https://github.com/home-assistant/android/blob/71ef9597890cd738f314bfac41e9db2fa5c6c65d/app/src/main/kotlin/io/homeassistant/companion/android/websocket/WebsocketManager.kt#L125

**Comment by engideer-tech** [comment]
@Rudd-O If the companion app were to implement UnifiedPush notifications via NTFY, would it still need its own Websocket connection (the one you were looking at)? If so, why? Sorry for the perhaps dumb question 😅

**Comment by synnack** [comment]
No it wouldn't and that's why the battery life would be so much better

**Comment by Bugsbane** [comment]
Why is this issue still closed? The reason given is no longer valid and surely it discourages contribution when the first thing potential contributors see is that it's "Closed as not planned".

We have everything we need and it's so very close!

**Comment by Kitt3120** [comment]
I'm actually surprised that this is not available yet

**Comment by mibofra** [comment]
@dshokouhi it seems it is time at least to open this issue again! It is the right time and community asks for the feature!

It will not pop up today or tomorrow, but at least reopen the issue to know it is taken into account, and acknowledge that possibile PRs are welcome!

**Comment by luneth** [comment]
@Rudd-O 
Actually, concerning the websocket ping internal, even if they "probably tried various timeouts with a shitty middlebox", 30 is a common value if the client is behind aggressive CGNAT.
I was playing along [trying to optimize](https://github.com/binwiederhier/ntfy-android/pull/113#issuecomment-3927658425) Ntfy's ping/pong, and I realized that:
- under perfect conditions, 300s interval client-sided pings were enough while behind 5g/LTE the client needed to ping every 45/60s, and could even need more depending of carrier
- Server ping is fine with stable connections but useless for mobile devices, even more so if they're async

Ntfy and Gotify use hardcoded 60s pings for the moment, not sure about sunup.

Even with logics using okhttp `onClosed` / `onFailure`, client ping is the only reliable way to check if the WS silently died.

Let's be honest Homeassistant wrecks the battery if updating sensors too often, just maintaining the websocket for timely notifications ~is not that bad~. (Edit: Nicholas might be right, I'll go through the code before making that assumption)

BUT having only one single application on your phone (UnifiedPush distributor - Ntfy/Nextcloud Push/Sunup...) maintaining a WS instead of a handful (let's say Homeassistant/Signal/matrix client/ etc...) Is huge battery wise, and well kind of the whole point of UP

**Comment by NicholasFlamy** [comment]
@luneth I'm suspicious that having Home Assistant holding a WebSocket has a lot of overhead. I'm not sure why but my hypothesis is that it's a bloated webview app (all webview apps are bloated) and that has something to do with it.
